source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -554,6 +614,369 @@ namespace aspect
+                std::vector<double> q_abs(4);
+                for (unsigned int i = 0; i < 4; i++)
+                  {
+                    q_abs[i] = std::abs(bigI[i] / tau[i]);
+                  }
+
+                // here we find the indices starting at the largest value and ending at the smallest value
+                // and assign them to special variables. Because all the variables are absolute values,
+                // we can set them to a negative value to ignore them. This should be faster then deleting
+                // the element, which would require allocation. (not tested)
+                for (unsigned int slip_system_i = 0; slip_system_i < 4; ++slip_system_i)
+                  {
+                    indices[slip_system_i] = std::distance(q_abs.begin(),max_element(q_abs.begin(), q_abs.end()));
+                    q_abs[indices[slip_system_i]] = -1;
+                  }
+
+                // compute the ordered beta vector, which is the relative slip rates of the active slips systems.
+                // Test whether the max element is not equal to zero.
+                Assert(bigI[indices[0]] != 0.0, ExcMessage("Internal error: bigI is zero."));
+                beta[indices[0]] = 1.0; // max q_abs, weak system (most deformation) "s=1"
+
+                const double ratio = tau[indices[0]]/bigI[indices[0]];
+                for (unsigned int slip_system_i = 1; slip_system_i < 4-1; ++slip_system_i)
+                  {
+                    const double beta_tmp = ratio * (bigI[indices[slip_system_i]]/tau[indices[slip_system_i]]);
+                    beta[indices[slip_system_i]] = beta_tmp * std::pow(std::abs(beta_tmp), stress_exponent-1);
+                  }
+                beta[indices.back()] = 0.0;
+
+                // Now compute the Crystal rate of deformation tensor.
+                for (unsigned int i = 0; i < 3; i++)
+                  {
+                    for (unsigned int j = 0; j < 3; j++)
+                      {
+                        G[i][j] = 2.0 * (beta[0] * rotation_matrix[0][i] * rotation_matrix[1][j]
+                                         + beta[1] * rotation_matrix[0][i] * rotation_matrix[2][j]
+                                         + beta[2] * rotation_matrix[2][i] * rotation_matrix[1][j]
+                                         + beta[3] * rotation_matrix[2][i] * rotation_matrix[0][j]);
+                      }
+                  }
+              }
+
+            // Now calculate the analytic solution to the deformation minimization problem
+            // compute gamma (equation 7, Kaminiski & Ribe, 2001)
+            double top = 0;
+            double bottom = 0;
+            for (unsigned int i = 0; i < 3; ++i)
+              {
+                // Following the Drex code, which differs from EPSL paper,
+                // which says gamma_nu depends on i+1: actually uses i+2
+                const unsigned int ip2 = (i==0) ? (i+2) : (i-1);
```

Comment:
it does indeed stand for "i plus 2", but since it wraps around from i=1 -> ip2 = 0 and i=2 -> ip2 = 1. I guess the best name would be `i_plus_2_wrapped_around`, but that is a bit long. I have changed it to `i_offset`. It is a bit generic, but maybe more clear. What do you think?